playbooks/kubernetes/app-resource-tuning.md

```diff
@@ -0,0 +1,111 @@
+
+- For less critical apps, set it to 1, or 2 if CPU request is set to greater than 1.
+
+### Memory Request
+- Observe memory usage over time. Set request to 1.5x that. For example, if a pod uses 512Mi consistently, set request to 768Mi.
+
+### Memory Limit
+- Set it to 2x usage. Example, if a pod uses 512Mi consistently, set it to 1Gi. Beware that there might be similar limit set on app-side. For example, Nodejs apps might be configured with `max_old_space_size` setting which should match the limit set in k8s.
+
+# Horizontal Pod Auto-Scaler (HPA)
+
+https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
+
+We scale our apps by scaling out, that is, by adding more instances. In k8s, each instance is a pod. Pods are managed by a `deployment`. HPA manages the pod count of a `deployment`. It adds or removes pods in response to changes in workload. It has several knobs:
+
+- `minReplicas` specifies the number of pods a deployment has at a minimum. Even if there's zero load, HPA still ensures there is that number of pods.
+- `maxReplicas` specifies the maximum number of pods a deployment can have. HPA caps the count at that, even if load is sky-high.
+- `targetCPUUtilizationPercentage` tells HPA to maintain pods' CPU utilization at that level. When utilization exeeds (or drops below) the target, HPA adds (or removes) pods, but pod count will be constrained by min/max settings.
+
+```
+apiVersion: autoscaling/v1
+kind: HorizontalPodAutoscaler
+metadata:
+  name: example-web
+...
+spec:
+  scaleTargetRef:
+    apiVersion: extensions/v1beta1
+    kind: Deployment
+    name: example-web
+  minReplicas: ???
+  maxReplicas: ???
+  targetCPUUtilizationPercentage: ???
+```
+## Recommendations
+
+### minReplicas
+- Set it to 2 for redundancy.
+
+### maxReplicas
+- Set it to as many as required to meet reasonable workload. We have unlimited capacity. Cluster Auto-Scaler will add EC2 instances to back up the pods. Cap it at a number beyond which you feel is unreasonable for the app.
+
+### targetCPUUtilizationPercentage
+- Set it to less than 100% so there is headroom for the pods to meet quick surges in workload, so that client requests are not dropped while HPA/CA add pods/EC2-instances which takes time. But don't set it so low that a lot of CPU sits idle. Try 70%-80%. For example, if set at 70%, pods' CPU utilization will be around 70%. When there's a surge in workload and HPA has added pods up to max, utilization will go beyond 70%.
+
+# Process of tuning those parameters.
+It takes iterations to get those parameters right. Generally, when launching an app, give it more resources (especially memory) than you think it needs, observe actual utilization over time, and dial down as needed. Here's a dashboard that helps monitoring an app's resource usage over time:

```

Comment:
🎉 encouragement is good here!

Reply:
+1

Since README is a public repo, we may want to add these links under a "Artsy staff..." subsection.

Reply:
Oh! Thanks @iskounen for pointing it out. Might it be more appropriate to place this doc in Notion?

https://www.notion.so/artsy/Engineering-Playbooks-b655fe54c1ce4b35af342c9ed9a489ae

Reply:
My rule of thumb is for support/troubleshooting/internal stuff to go into Notion, while generalizable content goes here. Since this is very coupled to our internal dashboards, I think it's closer to the former.

The general k8s utilization concepts and tips could always be published as its own playbook or blog post.

Reply:
Okay thanks. Created this doc in Notion: https://www.notion.so/artsy/Kubernetes-4d025f5c02c843f59fe4da3b743869b2
Closing PR.